src/client/mod.rs

```diff
@@ -638,4 +642,98 @@ impl Client {
         let mut state = self.state.write().await;
         state.last_heatbeat = Instant::now();
     }
+
+    async fn build_root_store(
+        root_ca_cert: Option<&Path>,
+    ) -> std::io::Result<rustls::RootCertStore> {
+        let mut roots = rustls::RootCertStore::empty();
+        let cert_bytes = std::fs::read(root_ca_cert.unwrap());
+
+        let root_cert_store = rustls_pemfile::certs(&mut cert_bytes.unwrap().as_ref()).unwrap();
+
+        root_cert_store
+            .iter()
+            .for_each(|cert| roots.add(&rustls::Certificate(cert.to_vec())).unwrap());
+        Ok(roots)
+    }
+
+    async fn build_client_certificates(
+        client_cert: &Path,
+    ) -> std::io::Result<Vec<rustls::Certificate>> {
+        let mut pem = BufReader::new(File::open(client_cert)?);
+        let certs = rustls_pemfile::certs(&mut pem)?;
+        let certs = certs.into_iter().map(rustls::Certificate);
+        Ok(certs.collect())
+    }
+
+    async fn build_client_private_keys(
+        client_private_key: &Path,
+    ) -> std::io::Result<Vec<PrivateKey>> {
+        let mut pem = BufReader::new(File::open(client_private_key)?);
+        let keys = rustls_pemfile::pkcs8_private_keys(&mut pem)?;
+        let keys = keys.into_iter().map(PrivateKey);
+        Ok(keys.collect())
+    }
+
+    async fn build_tls_client_configuration(
+        client_certificate_path: String,
+        roots: &rustls::RootCertStore,
+        broker: &ClientOptions,
+    ) -> Result<ClientConfig, ClientError> {
+        let config: ClientConfig;
+
+        if client_certificate_path.is_empty() {
+            config = ClientConfig::builder()
+                .with_safe_defaults()
+                .with_root_certificates(roots.clone())
+                .with_no_client_auth();
+        } else {
+            let client_certs = Self::build_client_certificates(Path::new(
+                &broker.tls.get_client_certificates_path(),
+            ))
+            .await?;
+            let client_keys =
+                Self::build_client_private_keys(Path::new(&broker.tls.get_client_keys_path()))
+                    .await?;
+            config = ClientConfig::builder()
+                .with_safe_defaults()
+                .with_root_certificates(roots.clone())
+                .with_client_auth_cert(client_certs, client_keys.into_iter().next().unwrap())
+                .unwrap();
+        }
+
+        Ok(config)
+    }
+
+    async fn build_tls_client_configuration_trusted() -> Result<ClientConfig, ClientError> {
+        mod danger {
```

Comment:
I would probably rename this to `build_tls_client_configuration_untrusted`

Reply:
Yeah indeed it seems like more appropriate fixed it :)